pallets/merkle-distributor/src/lib.rs

```diff
@@ -0,0 +1,328 @@
+    pub enum Error<T> {
+        /// Invalid metadata given.
+        BadDescription,
+        /// The id is not exist.
+        InvalidMerkleDistributorId,
+        /// The proof is invalid
+        MerkleVerifyFailed,
+        /// The reward is already distributed.
+        Claimed,
+        /// The reward is already charged.
+        Charged,
+    }
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(PhantomData<T>);
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// `create_merkle_distributor` will create a merkle distributor,
+        ///  which allow specified users claim asset.
+        ///
+        /// The dispatch origin for this call must be `Signed` by root.
+        ///
+        /// - `merkle_root`: The root of a merkle tree.
+        /// - `description`: About the purpose of this distribution.
+        /// - `distribute_currency`: The id of currency about this distribution.
+        /// - `distribute_amount`: The total currency amount of this distribution.
+        #[pallet::weight(T::WeightInfo::create_merkle_distributor())]
+        pub fn create_merkle_distributor(
+            origin: OriginFor<T>,
+            merkle_root: H256,
+            description: Vec<u8>,
+            distribute_currency: T::CurrencyId,
+            distribute_amount: T::Balance,
+        ) -> DispatchResult {
+            ensure_root(origin)?;
+
+            let merkle_distributor_id = Self::next_merkle_distributor_id();
+            let distribute_holder: AccountIdOf<T> =
+                T::PalletId::get().into_sub_account(merkle_distributor_id);
+
+            let description: BoundedVec<u8, T::StringLimit> = description
+                .try_into()
+                .map_err(|_| Error::<T>::BadDescription)?;
+
+            MerkleDistributorMetadata::<T>::insert(
+                merkle_distributor_id,
+                MerkleMetadata {
+                    merkle_root,
+                    description,
+                    distribute_currency,
+                    distribute_amount,
+                    distribute_holder,
+                },
+            );
+
+            Self::deposit_event(Event::<T>::Create(merkle_distributor_id, merkle_root, distribute_amount));
+            
+            Ok(())
+        }
+
+        /// `claim` Claim rewards through user information and merkle proof.
+        ///
+        /// - `merkle_distributor_id`: ID of a merkle distributor.
+        /// - `index`: The index of the merkle tree leaf.
+        /// - `account`: The owner's account of merkle proof.
+        /// - `merkle_proof`: The hashes with merkle tree leaf can get merkle tree root.
+        #[pallet::weight(T::WeightInfo::claim())]
+        #[transactional]
+        pub fn claim(
+            origin: OriginFor<T>,
+            merkle_distributor_id: T::MerkleDistributorId,
+            index: u32,
+            account: <T::Lookup as StaticLookup>::Source,
+            amount: u128,
+            merkle_proof: Vec<H256>,
```

Comment:
seems the proofs are generated offline in js code, just curious how to publish to end users, is there a rpc interface somewhere? would it be better to provide a rpc to generate merkle proof in runtime?

Reply:
seems merkle root is published by event already, only the proof part. IMHO github may not be a good place, a seperated rpc sevice to generate proof for each contributed user should be fine

Reply:
To generate a single user's proof, we need all users' data. If there are too many users, then the interface will not be able to bear. So, a centralized service should be best practice.